### ISSUE TITLE
ColorSpace 'alias' mechanism

### ISSUE THREAD

**Issue by zachlewis** [AcademySoftwareFoundation/OpenColorIO#1129]
*edit: See @doug-walker's proposal and discussion below
> [...] I'd like to propose an alternate solution which is to allow use of an "alias" key in a ColorSpace. This would allow the alternate name to be used anywhere the main name is used. This would include file rules but also command-line tools and other scenarios. Here is an example (the fourth line is the only difference from the current ACES config):
> 
> 
> ```
>   - !<ColorSpace>
>     name: ACES - ACEScct
>     family: ACES
>     alias: acescct
>     equalitygroup: ""
>     bitdepth: 32f
>     description: |
>       The ACEScct color space
>       
>       ACES Transform ID : ACEScsc.ACEScct_to_ACES
>     isdata: false
>     allocation: uniform
>     allocationvars: [-0.249136, 1.468]
>     to_reference: !<GroupTransform>
>       children:
>         - !<FileTransform> {src: ACEScct_to_linear.spi1d, interpolation: linear}
>         - !<MatrixTransform> {matrix: [0.695452, 0.140679, 0.163869, 0, 0.0447946, 0.859671, 0.0955343, 0, -0.00552588, 0.00402521, 1.0015, 0, 0, 0, 0, 1]}
> ```

Original post:
I would like to be able to repurpose the ColorSpace equalityGroup attribute to serve as ad-hoc colorspace 'aliases' where filename parsing is concerned. 

It's not uncommon for configs to contain copies of the same colorspace with alternate naming (and sometimes different families) --  E.g., "ACES - ACES2065-1", "aces", and "lin_ap0" are identical, and are found under different 'family paths'. 
Sometimes, a config will feature colorspaces with longer, human-friendly names prominently in interfaces and colorspace family hierarchy; and tucked away (or perhaps hidden in v2) are duplicate colorspaces with 'short names', often intended to serve as a filename 'color token', as is necessary for strict-parsing workflows. 

I propose either creating a new File Rule type for matching colorspace equalityGroups in a string, or updating the existing PathSearchRule to optionally search for equalityGroups as well. 

That's the primary proposal -- should be modest enough of a challenge for me to handle -- but it would be nice to reference colorspaces by their equality groups as shorthand for their actual names throughout OCIO; perhaps even have a place for them in menu interfaces, if wanted. 

Thoughts?

**Comment by sobotka** [comment]
+1 if there is a way to reduce the stanzas that define the actual transforms.

**Comment by doug-walker** [comment]
@zachlewis , thanks for raising this topic.  I've also had this on my list to bring up at a working group meeting because the use of aliases in the ACES config is less than ideal for a number of reasons:

1.  It requires duplicate ColorSpace definitions.  This bloats the config, is error-prone, and difficult to maintain.  For example, the equalitygroup is set on many of the aliases to point to the original but a lot of them do not, why?  Also, for those that do use the equalitygroup, there is no check to see if the redefined transforms actually do the same as the original.  This could be a difficult bug to track down, if there were an error.

2.  It adds a huge number of "duplicate" color spaces to application menus.  The new inactive_colorspaces feature could be used to remove these, but that was primarily intended to handle deprecated or internal colorspace definitions.  If the list grows to a hundred spaces (as it could with ACES), I think another mechanism would be helpful.  Also, one could make the case that a user would like to have the duplicates shown in menus in some cases, so that is another reason for wanting another mechanism than inactive_colorspaces.

3.  As you point out, the file rules would be nicer if you could set the color space for the rule using the friendly longer name but also have it pick up the shorter name automatically, without requiring duplicate rules.

Based on this, I'd like to propose an alternate solution which is to allow use of an "alias" key in a ColorSpace.  This would allow the alternate name to be used anywhere the main name is used.  This would include file rules but also command-line tools and other scenarios.  Here is an example (the fourth line is the only difference from the current ACES config):

```
  - !<ColorSpace>
    name: ACES - ACEScct
    family: ACES
    alias: acescct
    equalitygroup: ""
    bitdepth: 32f
    description: |
      The ACEScct color space
      
      ACES Transform ID : ACEScsc.ACEScct_to_ACES
    isdata: false
    allocation: uniform
    allocationvars: [-0.249136, 1.468]
    to_reference: !<GroupTransform>
      children:
        - !<FileTransform> {src: ACEScct_to_linear.spi1d, interpolation: linear}
        - !<MatrixTransform> {matrix: [0.695452, 0.140679, 0.163869, 0, 0.0447946, 0.859671, 0.0955343, 0, -0.00552588, 0.00402521, 1.0015, 0, 0, 0, 0, 1]}
```

**Comment by zachlewis** [comment]
Sorry equalitygroup lobby. Doug drives a really hard bargain. We'll get you next time.

**Comment by doug-walker** [comment]
This issue came up during the ACES Config working group meeting today so I wanted to add some additional points based on that discussion.

First of all, we felt it would be useful for the new alias attribute be allowed to have multiple strings as more than one alias may be desired.  For example, in the ACES Config we may want a "short name" alias for file paths (and tool use), as well as a "legacy name" to avoid breaking assets that reference names in the original ACES config (assuming that some of the names may change in the new version), as well as an "ACES Transform ID name" that might be useful for workflows involving ACES Metadata File.  Just to be clear, the config group has not decided to do any of these things (and there were different points of view expressed on the merits of each), I'm just stating them as examples that came up in the discussion of why it could be useful for the alias to allow more than a single string/name.

Second, we should be clear that the alias names would not show up by default in UI menus.  For example, if a DCC is generating its menus be calling getNumColorSpaces and then getColorSpaceNameByIndex to populate the menus, the alias names would not show up.  However, the API would allow a DCC to get the aliases, so if it wanted to make them visible to the user somehow (e.g. as a tooltip), it would be possible.  It just would not be the default.  The config author could also add the aliases to the description string for the color space to make them visible to users (although currently not many apps expose that either).  If a config author wants to ensure that a second menu item appears, they should add a duplicate color space (as the current ACES config does) rather than using the alias feature.

Third, it would be useful to have an API call that allows an app to ask for the "canonical" color space name for a given color space string.  For example, if a DCC is saving an asset that requires a color space name string (e.g. the color space of a texture), it may want to use the "canonical" name rather than an alias since it may be either more readable or more future-proof.  A use-case we discussed was if an asset was tagged with a legacy name, it would be nice if the DCC updated it to the canonical name before saving an updated version of the file to help migrate to the currently preferred name.

Fourth, there is potentially some overlap between the proposed alias feature and the new inactive_colorspaces feature.  Either could be used to handle the use-case where the name string of a color space has changed but the config does not want to break any assets that may reference the earlier name.  Of course, if there was a numerical difference then it may be necessary to keep both color space definitions and use the inactive_colorspaces.  Otherwise, if the number of name changes is large (e.g. if the ACES Config group changes the way the Input/Output/Utility prefixes are used) then the alias feature would certainly be the more convenient approach.

**Comment by sobotka** [comment]
Excellent work @doug-walker and all. Much needed, long overdue.

I originally thought your aliases design could account for multiple aliases, so it is great to see this take form.

On a similar thought train, can we get some discussion about mandatory roles that DCCs can lean on? It seems related to this idea, as a role is somewhat of an API-centric alias, and therefore, the defaults should be enforced and bolted down.

Excellent work!

**Comment by KelSolaar** [comment]
As discussed this morning, this is certainly on point and would be extremely useful to help the migration to config v2.

**Comment by maxnbk** [comment]
Was this issue solved/closed by #1241? Just trying to understand if there is more work here yet to do.

**Comment by zachlewis** [comment]
#1241 indeed resolves this. Closing.

Having experimented with this feature for months now, aliases are easily one of my favorite OCIO-2 improvements. The more you use them, the more you use them -- super convenient, _very_ powerful, lots of pipeline + ux potential, intuitively implemented in yaml and exposed via the API. A+++++ would buy from again!!
